discv5/discv5-theory.md

```diff
@@ -41,6 +40,141 @@ IP address and port.
 If the endpoint cannot be determined (e.g. when the NAT doesn't support 'full-cone'
 translation), implementations should omit IP address and UDP port from the record.
 
+## Sessions
+
+Discovery communication is encrypted and authenticated using session keys, established in
+the handshake.
+
+### Handshake
+
+Since every node participating in the network acts as both client and server, a handshake
+can be initiated by either side of communication at any time. In the following
+definitions, we assume that node A wishes to communicate with node B, e.g. to send a
+FINDNODE query.
+
+Node A must have a node record for node B and know B's node ID to communicate with it. If
+node A has session keys from prior communication, it encrypts its request with those keys.
+If no keys are known, it initiates the handshake by sending a packet with random content.
+
+    A -> B   FINDNODE encrypted with unknown key or random-packet
+
+Node B receives the initial packet, extracts the source node ID from the packet (see
+[encoding section]), and continues the handshake by responding with [WHOAREYOU]. The
+WHOAREYOU packet contains the id-nonce value to be signed by A as well as the highest
+known ENR sequence number of node A's record.
+
+    A <- B   WHOAREYOU including id-nonce, enr-seq
+
+Node A now knows that node B is alive and is ready to perform the handshake. The handshake
+proceeds by re-sending the original request message in a [handshake packet].
+
+The handshake packet includes an ephemeral public key in the cryptosystem used by B's
+identity scheme (e.g. an elliptic curve key on the secp256k1 curve if node B uses the "v4"
+scheme). This key is used to perform Diffie-Hellman key agreement with B's static public
+key and the session keys are derived from it using the HKDF key derivation function.
+
+    ephemeral-key    = random private key generated by node A
+    ephemeral-pubkey = public key corresponding to ephemeral-key
+    dest-pubkey      = public key of node B
+    secret           = ecdh(ephemeral-key, dest-pubkey)
+    info             = "discovery v5 key agreement" || node-id-A || node-id-B
+    prk              = HKDF-Extract(secret, id-nonce)
+
+    initiator-key, recipient-key = HKDF-Expand(prk, info)
+
+The handshake packet also contains a signature over `id-nonce` as well as node A's record
+if the local sequence number is higher than `enr-seq`. The signature proves that node A
+controls the identity key which signed the record and also prevents replay of the
+handshake.
+
+    id-nonce-input   = sha256("discovery-id-nonce" || id-nonce || ephemeral-pubkey)
```

Comment:
Assuming that `id-nonce` is a random value and `ephemeral-pubkey` is highly likely a random value, I am wondering does `"discovery-id-nonce"` constant give any valuable input to this `sha256` computation?

Reply:
It binds the signature to the specific purpose it is used for. Signatures are valuable and it is good practice to ensure that the input data is 'qualified', i.e. the input data says what the signature is supposed to mean. You can find many examples of this practice in other protocols as well.

Reply:
For example, this kind of thing is also used in TLS 1.3: https://www.davidwong.fr/tls13/#section-4.4.3

> The context string for a server signature is "TLS 1.3, server CertificateVerify". The context string for a client signature is "TLS 1.3, client CertificateVerify". It is used to provide separation between signatures made in different contexts, helping against potential cross-protocol attacks.

Reply:
Thanks for clarification. It makes sense in their context cause certificates are scarce resource wrt random values obtained from secure RNG, and different protocols may use the same certificates or keys. Anyway, following best practices does make a lot of sense. 

Do we want to mix a version, i.e. `1` in our case, into the context as well?

Reply:
Also, why not sign `id-nonce-input` with `ephemeral-secret-key` as `ephemeral-pubkey` goes alongside with the signature? In this case we could replace `ephemeral-pubkey` with `dest-node-id` for free which is probably a desirable thing.

Reply:
I don't understand the suggestion. The purpose of the ID signature is to prove that the initiator node actually has the private key that signed its ENR. The identity key of the initiator isn't used anywhere else in the handshake, so if we signed the ID nonce with the ephemeral key, the identity wouldn't be proven.

Reply:
Yes, you're right. The identity check is what I've been missing.